contribs/gmf/src/directives/print.js

```diff
@@ -0,0 +1,803 @@
+  if (done) {
+    // The report is ready. Open it by changing the window location.
+    window.location.href = this.ngeoPrint_.getReportUrl(ref);
+    this.resetPrintStates_();
+  } else {
+    // The report is not ready yet. Check again in 1s.
+    var that = this;
+    this.statusTimeoutPromise_ = this.$timeout_(function() {
+      that.getStatus_(ref);
+    }, 1000, false);
+  }
+};
+
+
+/**
+ * @param {!angular.$http.Response} resp Response.
+ * @private
+ */
+gmf.PrintController.prototype.handleCreateReportError_ = function(resp) {
+  this.resetPrintStates_(gmf.PrintState.ERROR_ON_REPORT);
+};
+
+
+/**
+ * @param {number} scale The scale to get the legend (for wms layers only).
+ * @return {Object?} Legend object for print report or null.
+ * @private
+ */
+gmf.PrintController.prototype.getLegend_ = function(scale) {
+  var legend = {'classes': []};
+  var classes, layerNames, layerName, icons;
+  var gettextCatalog = this.gettextCatalog_;
+
+  // Get layers from layertree only.
+  var dataLayerGroup = this.ngeoLayerHelper_.getGroupFromMap(this.map,
+      gmf.DATALAYERGROUP_NAME);
+  var layers = this.ngeoLayerHelper_.getFlatLayers(dataLayerGroup);
+
+  // For each visible layer in reverse order, get the legend url.
+  layers.reverse().forEach(function(layer) {
+    classes = [];
+    var source = layer.getSource();
+
+    if (layer.getVisible() && source !== undefined) {
+      // For WMTS layers.
+      if (layer instanceof ol.layer.Tile) {
+        layerName = layer.get('layerName');
+        icons = this.ngeoLayerHelper_.getWMTSLegendURL(layer);
+        // Don't add classes without legend url.
+        if (icons !== null) {
+          classes.push({
+            'name': gettextCatalog.getString('' + layerName),
+            'icons': [icons]
+          });
+        }
+      } else {
+        // For each name in a WMS layer.
+        layerNames = source.getParams()['LAYERS'].split(',');
+        layerNames.forEach(function(name) {
+          icons = this.ngeoLayerHelper_.getWMSLegendURL(source.getUrl(), name,
+              scale);
+          // Don't add classes without legend url or from layers without any
+          // active name.
+          if (icons !== null && name.length !== 0) {
+            classes.push({
+              'name': gettextCatalog.getString('' + name),
+              'icons': [icons]
+            });
+          }
+        }.bind(this));
+      }
+    }
+
+    // Add classes object only if it contains something.
+    if (classes.length > 0) {
+      legend['classes'].push({'classes': classes});
+    }
+
+  }.bind(this));
+
+  return legend['classes'].length > 0 ?  legend : null;
+};
+
+
+/**
+ * Set the current layout and update all fields with this new layout parameters.
+ * @param {string!} layoutName A layout name as existing in the list of
+ *     existing layouts.
+ * @export
+ */
+gmf.PrintController.prototype.setLayout = function(layoutName) {
+  var layout;
+  this.layouts_.forEach(function(l) {
```

Comment:
You could use [Array.find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) instead.